File: android/app/src/main/java/dev/envoix/app/discovery/BleVerificationInvitation.kt
package dev.envoix.app.discovery

import java.net.URI
import java.net.URLDecoder
import java.net.URLEncoder
import java.security.MessageDigest
import java.security.SecureRandom

/** Secret-free BLE locator. The six-digit PAKE input stays on the two screens. */
internal data class BleVerificationInvitation(
    val verificationCode: String,
    val privateInvitation: String,
    val publicOffer: String,
    val expiresAtEpochMs: Long,
) {
    companion object {
        const val URL_PREFIX = "envoix://ble/v1/"
        private const val LIFETIME_MS = 300_000L
        private val random = SecureRandom()

        fun create(
            broker: String,
            relay: String,
            nowEpochMs: Long = System.currentTimeMillis(),
        ): BleVerificationInvitation {
            val normalizedBroker = broker.trim()
            require(normalizedBroker.isNotEmpty()) { "A rendezvous broker is required" }
            val code = digits()
            var locator = digits()
            while (locator == code) locator = digits()
            val expiry = Math.addExact(nowEpochMs, LIFETIME_MS)
            val offer = url("ble", listOf("v1", locator), normalizedBroker, relay.trim(), expiry)
            return BleVerificationInvitation(
                code,
                privateUrl(locator, code, offer, normalizedBroker, relay.trim(), expiry),
                offer,
                expiry,
            )
        }

        fun resolve(
            publicOffer: String,
            verificationCode: String,
            nowEpochMs: Long = System.currentTimeMillis(),
        ): String? {
            if (!verificationCode.isDigits()) return null
            val offer = parse(publicOffer, nowEpochMs) ?: return null
            return privateUrl(
                offer.locator,
                verificationCode,
                publicOffer,
                offer.broker,
                offer.relay,
                offer.expiry,
            )
        }

        fun isPublicOffer(
            value: String,
            nowEpochMs: Long = System.currentTimeMillis(),
        ): Boolean = parse(value, nowEpochMs) != null

        private fun privateUrl(
            locator: String,
            code: String,
            publicOffer: String,
            broker: String,
            relay: String,
            expiry: Long,
        ): String {
            val secret =
                MessageDigest
                    .getInstance("SHA-256")
                    .digest("envoix BLE verification v1\u0000$publicOffer\u0000$code".toByteArray())
                    .take(4)
                    .joinToString("") { "%02x".format(it.toInt() and 0xff) }
            return url(
                "room",
                listOf("$locator-${secret.take(4)}-${secret.takeLast(4)}"),
                broker,
                relay,
                expiry,
            )
        }

        private fun url(
            host: String,
            path: List<String>,
            broker: String,
            relay: String,
            expiry: Long,
        ): String {
            val query =
                buildList {
                    add("broker=${broker.urlEncode()}")
                    if (relay.isNotEmpty()) add("relay=${relay.urlEncode()}")
                    add("expires=${expiry / 1_000L}")
                }.joinToString("&")
            return "envoix://$host/${path.joinToString("/")}?$query"
        }

        private fun parse(
            value: String,
            now: Long,
        ): Parsed? {
            if (value != value.trim() || value.toByteArray().size > 2_048) return null
            val uri = runCatching { URI(value) }.getOrNull() ?: return null
            val path = (uri.path ?: return null).split('/').filter(String::isNotEmpty)
            val rawFields = uri.rawQuery?.split('&') ?: return null
            if (rawFields.any { '=' !in it }) return null
            val fields =
                runCatching {
                    rawFields.map { field ->
                        val parts = field.split('=', limit = 2)
                        parts[0].urlDecode() to parts[1].urlDecode()
                    }
                }.getOrNull() ?: return null
            if (uri.scheme != "envoix" ||
                uri.host != "ble" ||
                uri.userInfo != null ||
                uri.port != -1 ||
                uri.fragment != null ||
                path.size != 2 ||
                path[0] != "v1" ||
                !path[1].isDigits() ||
                fields.any { it.first !in setOf("broker", "relay", "expires") }
            ) {
                return null
            }
            val broker = fields.one("broker")?.trim()?.takeIf(String::isNotEmpty) ?: return null
            val seconds = fields.one("expires")?.toLongOrNull() ?: return null
            if (fields.count { it.first == "relay" } > 1) return null
            val relay = fields.one("relay")?.trim().orEmpty()
            val expiry = runCatching { Math.multiplyExact(seconds, 1_000L) }.getOrNull() ?: return null
            if (broker.toByteArray().size > 1_024 ||
                relay.toByteArray().size > 1_024 ||
                expiry <= now ||
                expiry - now > LIFETIME_MS * 2
            ) {
                return null
            }
            return Parsed(path[1], broker, relay, expiry)
        }

        private fun List<Pair<String, String>>.one(name: String): String? =
            filter { it.first == name }.takeIf { it.size == 1 }?.single()?.second

        private fun String.urlEncode(): String = URLEncoder.encode(this, Charsets.UTF_8.name())

        private fun String.urlDecode(): String = URLDecoder.decode(this, Charsets.UTF_8.name())

        private fun digits(): String = random.nextInt(1_000_000).toString().padStart(6, '0')

        private fun String.isDigits(): Boolean = length == 6 && all { it in '0'..'9' }

        private data class Parsed(
            val locator: String,
            val broker: String,
            val relay: String,
            val expiry: Long,
        )
    }
}

File: apps/envoix-apple/Sources/BleVerificationInvitation.swift
import Foundation
import CryptoKit

/// Secret-free BLE locator. The six-digit PAKE input stays on the two screens.
struct BleVerificationInvitation: Equatable, Identifiable {
    static let urlPrefix = "envoix://ble/v1/"
    private static let lifetime: TimeInterval = 300

    let verificationCode: String
    let privateInvitation: String
    let publicOffer: String
    let expiresAt: Date
    var id: String { publicOffer }

    static func make(broker: String, relay: String, now: Date = Date()) throws -> Self {
        let configuredBroker = broker.trimmed
        let broker = configuredBroker.isEmpty ? defaultRendezvousBroker : configuredBroker
        let configuredRelay = relay.trimmed
        let relay = configuredRelay.isEmpty ? defaultRelayURL : configuredRelay
        let code = digits()
        var locator = digits()
        while locator == code { locator = digits() }
        let expiry = now.addingTimeInterval(lifetime)
        let offer = url("ble", ["v1", locator], broker, relay, expiry)
        return Self(
            verificationCode: code,
            privateInvitation: privateURL(locator, code, offer, broker, relay, expiry),
            publicOffer: offer,
            expiresAt: expiry
        )
    }

    static func resolve(
        publicOffer: String,
        verificationCode: String,
        now: Date = Date()
    ) -> String? {
        guard isDigits(verificationCode), let offer = parse(publicOffer, now: now) else { return nil }
        return privateURL(
            offer.locator,
            verificationCode,
            publicOffer,
            offer.broker,
            offer.relay,
            offer.expiry
        )
    }

    static func isPublicOffer(_ value: String, now: Date = Date()) -> Bool {
        parse(value, now: now) != nil
    }

    private static func privateURL(
        _ locator: String,
        _ code: String,
        _ publicOffer: String,
        _ broker: String,
        _ relay: String,
        _ expiry: Date
    ) -> String {
        let material = Data("envoix BLE verification v1\0\(publicOffer)\0\(code)".utf8)
        let secret = SHA256.hash(data: material).prefix(4)
            .map { String(format: "%02x", $0) }
            .joined()
        return url(
            "room",
            ["\(locator)-\(secret.prefix(4))-\(secret.suffix(4))"],
            broker,
            relay,
            expiry
        )
    }

    private static func url(
        _ host: String,
        _ path: [String],
        _ broker: String,
        _ relay: String,
        _ expiry: Date
    ) -> String {
        var value = URLComponents()
        value.scheme = "envoix"
        value.host = host
        value.path = "/" + path.joined(separator: "/")
        value.queryItems = [URLQueryItem(name: "broker", value: broker)]
        if !relay.isEmpty { value.queryItems?.append(URLQueryItem(name: "relay", value: relay)) }
        value.queryItems?.append(URLQueryItem(
            name: "expires",
            value: String(Int64(expiry.timeIntervalSince1970))
        ))
        return value.string!
    }

    private static func parse(
        _ input: String,
        now: Date
    ) -> (locator: String, broker: String, relay: String, expiry: Date)? {
        guard input == input.trimmed, input.utf8.count <= 2_048,
              let value = URLComponents(string: input),
              value.scheme == "envoix", value.host == "ble", value.fragment == nil,
              value.user == nil, value.password == nil, value.port == nil,
              let rawItems = value.percentEncodedQuery?.split(
                  separator: "&",
                  omittingEmptySubsequences: false
              ), rawItems.allSatisfy({ $0.contains("=") }) else { return nil }
        let path = value.path.split(separator: "/")
        let items = value.queryItems ?? []
        guard path.count == 2, path[0] == "v1", isDigits(String(path[1])),
              items.allSatisfy({ ["broker", "relay", "expires"].contains($0.name) }),
              let broker = one("broker", items)?.trimmed, !broker.isEmpty,
              let seconds = one("expires", items).flatMap(Int64.init),
              items.filter({ $0.name == "relay" }).count <= 1 else { return nil }
        let expiry = Date(timeIntervalSince1970: TimeInterval(seconds))
        let relay = one("relay", items)?.trimmed ?? ""
        guard broker.utf8.count <= 1_024, relay.utf8.count <= 1_024,
              expiry > now, expiry.timeIntervalSince(now) <= lifetime * 2 else { return nil }
        return (String(path[1]), broker, relay, expiry)
    }

    private static func one(_ name: String, _ items: [URLQueryItem]) -> String? {
        let matches = items.filter { $0.name == name }
        return matches.count == 1 ? matches[0].value : nil
    }

    private static func digits() -> String { String(format: "%06d", Int.random(in: 0..<1_000_000)) }
    private static func isDigits(_ value: String) -> Bool {
        value.utf8.count == 6 && value.utf8.allSatisfy { (48...57).contains($0) }
    }
}
